src/transformers/modeling_attn_mask_utils.py

```diff
@@ -234,6 +234,59 @@ def _unmask_unattended(
 
         return expanded_mask.mul(~torch.all(expanded_mask == min_dtype, dim=-1, keepdim=True))
 
+    @staticmethod
+    def _ignore_causal_mask_sdpa(
+        attention_mask: Optional[torch.Tensor],
+        inputs_embeds: torch.Tensor,
+        past_key_values_length: int,
+    ) -> bool:
+        """
+        Detects whether the attention_mask can be ignored in case we use PyTorch's SDPA.
+
+        In case no token is masked in the `attention_mask` argument, if `query_length == 1` or
+        `key_value_length == query_length`, we rather rely on SDPA `is_causal` argument to use causal/non-causal masks,
+        allowing to dispatch to the flash attention kernel (that can otherwise not be used if a custom `attn_mask` is passed).
+        """
+
+        batch_size, query_length = inputs_embeds.shape[0], inputs_embeds.shape[1]
+        key_value_length = query_length + past_key_values_length
+
+        is_tracing = (
+            torch.jit.is_tracing()
+            or isinstance(inputs_embeds, torch.fx.Proxy)
+            or (hasattr(torch, "_dynamo") and torch._dynamo.is_compiling())
+        )
+
+        ignore_causal_mask = False
+
+        if attention_mask is None:
+            # TODO: When tracing with TorchDynamo with fullgraph=True, the model is recompiled depending on the input shape, thus SDPA's `is_causal` argument is rightfully updated (see https://gist.github.com/fxmarty/1313f39037fc1c112508989628c57363). However, when using `torch.export` or
+            # or `torch.onnx.dynamo_export`, we must pass an example input, and `is_causal` behavior is hard-coded. If a user exports a model with q_len > 1, the exported model will hard-code `is_causal=True` which is in general wrong (see https://github.com/pytorch/pytorch/issues/108108).
+            # Thus, we currently can NOT set `ignore_causal_mask = True` here. We would need a `torch._dynamo.is_exporting()` flag.
+            #
+            # Besides, jit.trace can not handle the `q_len > 1` condition for `is_causal` (`TypeError: scaled_dot_product_attention(): argument 'is_causal' must be bool, not Tensor`).
+            ignore_causal_mask = not is_tracing
```

Comment:
@warner-benjamin I'm quite sure there is no way around this currently. We may revisit this with https://github.com/pytorch/pytorch/pull/114823 if that helps.

But I get it that we should have a path where users don't use `attention_mask` input, and use torch.compile WITHOUT `torch.export`/`torch.onnx.dynamo_export`.

Reply:
With my proposed modification to `SdpaAttention.forward`, I think it would be possible to check here for `self.training` to dispatch to FA2 and Efficient if in eval? Otherwise, a config option to force SDPA FA2 would work.

Reply:
@fxmarty I've been unable to trigger the full graph invalidation as you've shown. Could you share code to reproduce it?

I've tried my proposed changes with `torch.export` and `torch.compile` with `fullgraph=True`, `dynamic=True`, and both `fullgraph=True, dynamic=True` with PyTorch 2.2.2 and PyTorch 2.1.2 (minus export since it's a 2.2 feature). I have been unsuccessful in triggering any recompilations (outside of different input shapes without `dynamic=True`, but that's too be expected).

> What I am saying is that we want to avoid hard-coding `is_causal`. As no `torch._dynamo.is_exporting()` is not available (I wish it was, as well as `torch._dynamo.is_fullgraph_compiling()` [pytorch/pytorch#120400](https://github.com/pytorch/pytorch/pull/120400)), my safe bet is to always use `attn_mask` in case torch.compile is used.

I'm not following why we cannot use the FA2 kernel during compiled training. We could dispatch via `self.training` to FA2/`attn_mask`, add a config option to force FA2/`attn_mask`, or both.

As it currently stands, someone testing their model in eager mode would use the memory efficient FA2 kernel, and then when switching to `torch.compile` for additional memory savings and speed, their training would instead use more memory due to silently switching the kernel from FA2 to efficient.

Reply:
> @fxmarty I've been unable to trigger the full graph invalidation as you've shown. Could you share code to reproduce it?

Sure, let me do next week.

> I'm not following why we cannot use the FA2 kernel during compiled training. We could dispatch via self.training to FA2/attn_mask, add a config option to force FA2/attn_mask, or both.

Sure, `self.training` seem to be a good option

Reply:
Thanks. Will fix compile+FA2 in a new PR once this one is merged.

Reply:
Thank you for your thorough review!